### PR TITLE
[TUE] Enh: better handling of frame range update (RDEV-3011)

### PR DESCRIPTION
**Pull request by gst**
**Purpose of the PR / What does this do?**   
 
Enhance handling of set framerange (in nuke but not only).

+ All : Don't ask confirmation to update frame range from shotgun if it's already the same than currently.
+ nuke : just slightly better dialog messages for locked or not locked cases.

**Type of feedback wanted**    usual
 
**Where should the reviewer start looking at?**   
 
+ https://github.com/rodeofx/tk-multi-setframerange/compare/master_rdo...rodeofx:dev_nuke_framerange_RDEV-3011?expand=1#diff-3f41e546893dc64b71aaacad12cad815R59

**Potential risks of this change, if any.**   Should be None.
 
**Related PR** (forgot about it) : 

https://bitbucket.org/rodeofx/tk_config_rdo_default/pull-requests/319/to-use-last-update-of-tk-multi/diff

### CODE REVIEW

**Comment by gst** [comment]
NB: can be tested with tank config "framerange" on rom show : 

`nuke rom +c framerange`

File: app.py [hunk 1]
```diff
@@ -185,17 +186,37 @@ def set_frame_range(self, engine, in_frame, out_frame):
...
+            locked = lock_range.value()
+            change_msg = 'from %s-%s to %s-%s' % (prev_in, prev_out, in_frame, out_frame)
             if locked:
-                nuke.root()["lock_range"].setValue(False)
+                ret = nuke.ask("Frame range is currently locked,\n"
+                               "OK to update frame range %s ?" % change_msg)
+                if not ret:
+                    return
+            else:
+                p = nuke.Panel("Shall lock ?")
```
Comment (JeanChristopheMorinRodeoFX) [inline app.py]: I would prefer if you keep using Qt stuff directly and not Nuke stuff. Using Qt stuff will make it clear that it comes from this app and Tank as Tank applies some styling to widgets/dialogs, ansd so we can easily know the pop-up is coming from Tank apps/engines.

File: app.py [hunk 2]
```diff
@@ -185,17 +186,37 @@ def set_frame_range(self, engine, in_frame, out_frame):
...
+                p.addMultilineTextInput("info",
+                                        "Frame range is going to be updated \n\n"
+                                        "%s,\n\n"
+                                        "please choose what to do.." % change_msg)
+                p.addButton('Reject')  # 0
+                p.addButton('Accept')  # 1
+                p.addButton('Accept but then LOCK the frame range')  # 2
```
Comment (JeanChristopheMorinRodeoFX) [inline app.py]: Minor: To make it shorter: `Accept and lock`. I don't think it is necessary to re-mention `frame range` as this dialog is already talking about frame range.

File: app.py [hunk 3]
```diff
@@ -185,17 +186,37 @@ def set_frame_range(self, engine, in_frame, out_frame):
...
+            root = nuke.root()
+            lock_range = root["lock_range"]
             # unlock
-            locked = nuke.root()["lock_range"].value()
+            locked = lock_range.value()
+            change_msg = 'from %s-%s to %s-%s' % (prev_in, prev_out, in_frame, out_frame)
             if locked:
-                nuke.root()["lock_range"].setValue(False)
+                ret = nuke.ask("Frame range is currently locked,\n"
+                               "OK to update frame range %s ?" % change_msg)
```
Comment (JeanChristopheMorinRodeoFX) [inline app.py]: Talking from a Tank perspective, `Do you want to update frame range %s` would be more tanky than `OK`, as they normally don't use such phrases.

**Comment by kevinsallee** [comment]
looks good to me :+1:

File: app.py [hunk 4]
```diff
@@ -52,33 +51,36 @@ def run_app(self):
         (current_in, current_out) = self.get_current_frame_range(self.engine.name)
 
         if new_in is None or new_out is None:
-            message =  "Shotgun has not yet been populated with \n"
-            message += "in and out frame data for this Shot."
+            message = ("Shotgun has not yet been populated with \n"
+                       "in and out frame data for this Shot.")
```
Comment (JeanChristopheMorinRodeoFX) [inline app.py]: I know the way you did is better, but changing stuff like this will make it harder to later upgrade with a version coming from Shotgun. We need to avoid unnecessary changes as much as possible, when possible. The more changes we do, and the harder it is to merge Shotgun's master branch into ours.

File: app.py [hunk 5]
```diff
@@ -185,17 +186,37 @@ def set_frame_range(self, engine, in_frame, out_frame):
...
+                ret = p.show()
+                if ret == 0:  # reject
+                    return
+                if ret == 2:  # accept and lock
+                    locked = True
+            lock_range.setValue(False)
             # set values
-            nuke.root()["first_frame"].setValue(in_frame)
-            nuke.root()["last_frame"].setValue(out_frame)
+            root["first_frame"].setValue(in_frame)
```
Comment (JeanChristopheMorinRodeoFX) [inline app.py]: As mentioned upper in the PR, I would prefer if we avoid unnecessary changes when possible. Same for line 26 and 219.

**Review by JeanChristopheMorinRodeoFX — CHANGES_REQUESTED** [review]

**Comment by JeanChristopheMorinRodeoFX** [comment]
Except my little stylish comments, I have to say: Good job!

File: app.py (hunk 3 above)
Comment (darkvertex) [reply app.py]: I agree.

File: app.py (hunk 1 above)
Comment (darkvertex) [reply app.py]: I think this is a good idea, too.

**Review by darkvertex — CHANGES_REQUESTED** [review]

File: app.py (hunk 4 above)
Comment (frenchdog) [reply app.py]: I agree with @JeanChristopheMorin but I'm curious to know how far we are already from the Shotgun code. If we are already very customized, then we don't care. If we are currently matching what is comming from Shotgun, then I won't approve such PR.

Comment (JeanChristopheMorinRodeoFX) [reply app.py]: We are not really far from their code. ATM, the only thing that we added is the XSI support and we did one or two hotfix (like force variables to be int when setting frame range in nuke)

Comment (frenchdog) [reply app.py]: In this case @gst , lets not pollute existing code not related to the frame range. Thanks!

Comment (gst) [reply app.py]: ok (holycrap)

File: app.py (hunk 3 above)
Comment (gst) [reply app.py]: ..

File: app.py (hunk 1 above)
Comment (gst) [reply app.py]: ..

File: app.py (hunk 5 above)
Comment (gst) [reply app.py]: holy..

**Comment by gst** [comment]
comments handled.

**Review by darkvertex — APPROVED** [review]

File: app.py [hunk 6]
```diff
@@ -185,17 +186,39 @@ def set_frame_range(self, engine, in_frame, out_frame):
...
+            lock_range = root["lock_range"]
             # unlock
-            locked = nuke.root()["lock_range"].value()
+            locked = lock_range.value()
+            change_msg = 'from %s-%s to %s-%s' % (prev_in, prev_out, in_frame, out_frame)
             if locked:
-                nuke.root()["lock_range"].setValue(False)
+                box = QtGui.QMessageBox()
+                box.setText("Frame range is currently locked,\n"
+                            "Do you want to still update it %s ?" % change_msg)
```
Comment (JeanChristopheMorinRodeoFX) [inline app.py]: `Do you still want to update it` from what people say in the developpers channel in Slack.

**Review by JeanChristopheMorinRodeoFX — APPROVED** [review]

**Comment by gst** [comment]
what makes me laugh / questino, is that the knob "lock_range" seems basically of no (or very little) use to me: 

if you leave it to True and update the frame_range :  it's well updated.. :D

File: app.py [hunk 7]
```diff
@@ -185,17 +192,21 @@ def set_frame_range(self, engine, in_frame, out_frame):
...
+                return
             # set values
-            nuke.root()["first_frame"].setValue(in_frame)
-            nuke.root()["last_frame"].setValue(out_frame)
-            # and lock again
-            if locked:
-                nuke.root()["lock_range"].setValue(True)
+            root['first_frame'].setValue(in_frame)
+            root['last_frame'].setValue(out_frame)
+            root['lock_range'].setValue(True)
```
Comment (JeanChristopheMorinRodeoFX) [inline app.py:209]: So you now always lock?

Comment (gst) [reply app.py:209]: yes, requested by Patrick David, after consensus with his team.

I know this gives them less control and I explained him it was better to keep the possibility to lock or not but he disregarded that.

File: app.py [hunk 8]
```diff
@@ -34,16 +35,16 @@ def init_app(self):
                                  "a current Shot, current Asset etc). This app requires "
                                  "an entity as part of the context in order to work.")
 
-        self.engine.register_command("Sync Frame Range with Shotgun", self.run_app)
+        self.engine.register_command("Sync Frame Range with Shotgun",
+                                     lambda: self.run_app(from_menu=True))
```
Comment (JeanChristopheMorinRodeoFX) [inline app.py:39]: Excuse me for this, but what is the lambda doing here? Not sure to fully understand.

Comment (gst) [reply app.py:39]: heuh, the same as previously, but it calls `run_app` with `from_menu` set to `True` ?

File: app.py [hunk 9]
```diff
@@ -161,7 +168,7 @@ def get_current_frame_range(self, engine):
 
         return (current_in, current_out)
 
-    def set_frame_range(self, engine, in_frame, out_frame):
+    def set_frame_range(self, engine, in_frame, out_frame, prev_in, prev_out):
```
Comment (mfe) [inline app.py:171]: I would have try to not break the API by having new params as kwargs and handle the case where they are not defined (especially since they seems to be used in a text filed only ?)  But I guess you check all set_frame_range calls ?

File: app.py (hunk 7 above)
Comment (mfe) [reply app.py:209]: Was wondering this yesterday, thanks for answering :+1:

**Review by mfe — APPROVED** [review]
Minor comment that might not be pertinent, so LGTM especially since Pat validated the changes.

File: app.py (hunk 9 above)
Comment (gst) [reply app.py:171]: `set_frame_range` is hopefully only used in this app.. or otherwise we would be abusing how the tank apps system work **if I'm not mistaken**   ,  @JeanChristopheMorin  ??

Comment (JeanChristopheMorinRodeoFX) [reply app.py:171]: Yes. If we want to run an app, we normally need to use `app.run_app`. And normally, we "should" not use other apps methods (Note the "normally")

Comment (gst) [reply app.py:171]: yeah.. ;) but I really hope we are not abusing that anywhere.. or else I'll kill them all :)
